File: packages/common/erp-common-ms-utils-py/gmahechas_erp_common_ms_utils_py/helpers/vault.py
from hvac import Client
from os import environ
from requests import post


class Vault:
    __instance = None

    def __init__(self, url, token):
        if Vault.__instance is not None:
            raise Exception("Config already initialized")
        else:
            Vault.__instance = Client(url=url, token=token)

    @staticmethod
    def get_instance():
        if Vault.__instance is None:
            raise Exception("Vault not initialized")
        return Vault.__instance

    @staticmethod
    def auth(url, role_id, secret_id):
        data = post(f"{url}/v1/auth/approle/login", json={'role_id': role_id, 'secret_id': secret_id})
        return data.json()

    @staticmethod
    def read(path):
        vault_secrets = Vault.__instance.secrets.kv.v2.read_secret_version(mount_point='kv', path=path)['data'][
            'data']
        return vault_secrets


def init_vault():
    vault_url = environ["VAULT_URL"]
    vault_role_id = environ["VAULT_ROLE_ID"]
    vault_secret_id = environ["VAULT_SECRET_ID"]

    data = Vault.auth(vault_url, vault_role_id, vault_secret_id)
    client_token = data['auth']['client_token']
    vault = Vault(url=vault_url, token=client_token)

File: packages/common/erp-common-ms-utils-go/helpers/vault.go
package helpers

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"

	vault "github.com/hashicorp/vault/api"
	"github.com/spf13/viper"
)

type VaultParameters struct {
	vaultUrl      string
	vaultRoleId   string
	vaultSecretId string
}

type Vault struct {
	*vault.Client
}

func (vault *Vault) Read(path string) (map[string]interface{}, error) {
	data, err := vault.Logical().Read(path)
	if err != nil {
		fmt.Println(err)
		return nil, err
	}
	return data.Data, nil
}

func NewVault(v *Vault, token string, parameters VaultParameters) (*Vault, error) {
	config := vault.DefaultConfig()
	config.Address = parameters.vaultUrl
	client, _ := vault.NewClient(config)
	*v = Vault{client}
	(*v).SetToken(token)
	return v, nil
}

var MyVault Vault

func InitVault() {
	vaultUrl := viper.GetString("VAULT_URL")
	vaultRoleId := viper.GetString("VAULT_ROLE_ID")
	vaultSecretId := viper.GetString("VAULT_SECRET_ID")

	jsonData, _ := json.Marshal(map[string]string{"role_id": vaultRoleId, "secret_id": vaultSecretId})
	data, _ := http.Post(vaultUrl+"/v1/auth/approle/login", "application/json", bytes.NewBuffer(jsonData))
	var res map[string]interface{}
	json.NewDecoder(data.Body).Decode(&res)
	token := res["auth"].(map[string]interface{})["client_token"].(string)

	NewVault(&MyVault, token, VaultParameters{vaultUrl, vaultRoleId, token})
}
